### PR TITLE
Lightwallet Daemon Mode, Watchonly Daemon Mode Inital Code

### PR DESCRIPTION
**Pull request by blondfrogs**
This is the first PR which includes the needed code for lightwallet ringct and stealth transactions. 

the flag -lightwallet is added. If this flag is added to the veil.conf or as a startup flag this daemon will only act as an RPC server. It will not sync blocks, it will not connect to peers. This daemon will be used for building the ringct and stealth transactions. 

the flag -watchonly is added. If this flag is added to the veil.conf or as a startup flag this daemon will act like a fullnode, however will not have a functional wallet. So we don't send veil to this wallets addresses. This daemon will act like the backend server which light wallet clients will connect to to retrieve information about their stealth addresses. 

This PR has been tested on testnet as a normal veil node, lightwallet, and watchonly node. The node functions correctly in all states. 

If you want to know more information on how lightwallet transactions work please visit the new README file in doc/lightwallet.md

### CODE REVIEW

File: src/core_io.h [hunk 1]
```diff
@@ -38,4 +42,9 @@ void ScriptPubKeyToUniv(const CScript& scriptPubKey, UniValue& out, bool fInclud
 void ScriptToUniv(const CScript& script, UniValue& out, bool include_address);
 void TxToUniv(const CTransaction& tx, const uint256& hashBlock, const std::vector<std::vector<COutPoint>>& vTxRingCtInputs, UniValue& entry, bool include_hex = true, int serialize_flags = 0);
 
+void OutputToJSON(const uint256 &txid, const int& i,const CTxOutBase *baseOut, UniValue &entry, bool isCoinBase = false);
+void RingCTOutputToJSON(const uint256& txid, const int& i, const int64_t& ringctIndex, const CTxOutRingCT& ringctOut, UniValue &entry);
```
Comment (Zannick) [inline src/core_io.h:46]: There's really no point in having const int& and const int64_t&, just make them regular int and int64_t (pass by value).

File: src/init.cpp [hunk 2]
```diff
@@ -274,6 +280,18 @@ void Shutdown()
         DumpMempool();
     }
 
+    if (gArgs.GetBoolArg("-watchonly", false)) {
+        if (!FlushWatchOnlyAddresses()) {
+            error("Fail to write watchonly addresses to database.");
```
Comment (Zannick) [inline src/init.cpp:285]:
```suggestion
            error("Failed to write watchonly addresses to database.");
```

File: src/init.cpp [hunk 3]
```diff
@@ -274,6 +280,18 @@ void Shutdown()
...
+        }
+
+        {
+            LOCK(cs_watchonly);
+            // Ran into memory problems if this map wasn't cleared before shutdown.
```
Comment (Zannick) [inline src/init.cpp:290]: That sounds like there may be another issue, perhaps this is a global? This would be better encapsulated somewhere in a destructor if not.

File: src/miner.cpp [hunk 4]
```diff
@@ -908,7 +908,7 @@ void BitcoinMiner(std::shared_ptr<CReserveScript> coinbaseScript, bool fProofOfS
 #ifdef ENABLE_WALLET
         if (enablewallet && fProofOfStake) {
             if (IsInitialBlockDownload()) {
-                UninterruptibleSleep(std::chrono::milliseconds{60000});
+                UninterruptibleSleep(std::chrono::milliseconds{5000});
```
Comment (Zannick) [inline src/miner.cpp:911]: Were these changes from testing or something? Do we need them?

File: src/txdb.h [hunk 5]
```diff
@@ -23,6 +23,7 @@
 class CBlockIndex;
 class CCoinsViewDBCursor;
 class uint256;
+class CBitcoinAddress;
```
Comment (Zannick) [inline src/txdb.h:26]: imo if it's not used in the header file we shouldn't declare it.

File: src/validation.cpp [hunk 6]
```diff
@@ -3017,9 +3031,17 @@ bool static FlushStateToDisk(const CChainParams& chainparams, CValidationState &
                     return AbortNode(state, "Failed to write to block index database");
                 }
             }
+
+            if (gArgs.GetBoolArg("-watchonly", false)) {
+                if (!FlushWatchOnlyAddresses()) {
+                    return AbortNode(state, "Fail to write watchonly addresses to database.");
```
Comment (Zannick) [inline src/validation.cpp:3037]:
```suggestion
                    return AbortNode(state, "Failed to write watchonly addresses to database.");
```

File: src/miner.cpp [hunk 7]
```diff
@@ -1179,6 +1180,12 @@ void BitcoinRandomXMiner(std::shared_ptr<CReserveScript> coinbaseScript, int vm_
                 randomx_calculate_hash(vecRandomXVM[vm_index], &nHeaderHash, sizeof uint256(), hash);
 
                 uint256 nHash = RandomXHashToUint256(hash);
+
+                // Bypass regtest check, actually allows us to generate blocks in regtest mode instantly
+                if (Params().NetworkIDString() == "regtest") {
```
Comment (Zannick) [inline src/miner.cpp:1185]: I don't know that we want to commit this.

File: src/veil/ringct/anonwallet.cpp [hunk 8]
```diff
@@ -636,14 +636,19 @@ isminetype AnonWallet::IsMine(const CTxOutBase *txout) const
         case OUTPUT_RINGCT:
         {
             CKeyID keyID = ((CTxOutRingCT*) txout)->pk.GetID();
-            if (mapKeyPaths.count(keyID))
+            if (mapKeyPaths.count(keyID)) {
+                CKeyID keyID = ((CTxOutRingCT*) txout)->pk.GetID();
```
Comment (Zannick) [inline src/veil/ringct/anonwallet.cpp:640]: Accidentally duplicated line? Doesn't seem like it should have an effect.

File: src/veil/ringct/anonwallet.cpp [hunk 9]
```diff
@@ -3136,6 +3147,99 @@ bool AnonWallet::PickHidingOutputs(std::vector<std::vector<int64_t> > &vMI, size
     return true;
 }
 
+
+bool AnonWallet::GetRandomHidingOutputs(size_t nInputSize, size_t nRingSize, std::set<int64_t> &setHave, std::vector<CLightWalletAnonOutputData>& randomoutputs, std::string &sError)
```
Comment (Zannick) [inline src/veil/ringct/anonwallet.cpp:3151]: Again it would be ideal to have some shared code for use by these two functions, rather than duplicate a lot of tricky code. If possible.

File: src/veil/ringct/anonwallet.cpp [hunk 10]
```diff
@@ -4684,8 +4788,35 @@ bool AnonWallet::AddKeyToParent(const CKey& keySharedSecret)
     return fSuccess;
 }
 
+void AnonWallet::GetAllScanKeys(std::vector<CStealthAddress>& vStealthAddresses) {
+    LOCK(pwalletParent->cs_wallet);
+    int i = 0;
+    std::set<std::string> seenAddresses;
+
+    for (auto stealth: mapStealthDestinations) {
```
Comment (Zannick) [inline src/veil/ringct/anonwallet.cpp:4796]:
```suggestion
    for (const auto& stealth: mapStealthDestinations) {
```
if these are usable as const references.

File: src/veil/ringct/lightwallet.h [hunk 11]
```diff
@@ -0,0 +1,63 @@
+//
+// Created by main on 25/2/22.
```
Comment (Zannick) [inline src/veil/ringct/lightwallet.h:2]: Clean up these comments. Replace with a standard copyright header seen in other files.

File: src/veil/ringct/lightwallet.cpp [hunk 12]
```diff
@@ -0,0 +1,1890 @@
+//
+// Created by main on 25/2/22.
```
Comment (Zannick) [inline src/veil/ringct/lightwallet.cpp:2]: Replace with a standard copyright comment seen in other files.

File: src/veil/ringct/lightwallet.cpp [hunk 13]
```diff
@@ -0,0 +1,1890 @@
...
+#include <veil/ringct/temprecipient.h>
+#include <secp256k1_mlsag.h>
+#include <veil/ringct/blind.h>
+#include <wallet/coincontrol.h>
+#include <veil/ringct/anonwallet.h>
+#include <policy/policy.h>
+#include <core_io.h>
+
+
+std::string TestBuildWalletTransaction(int nRandomInt) {
```
Comment (Zannick) [inline src/veil/ringct/lightwallet.cpp:20]: This should be in a separate test file, under tests/ or something, not in the main binary.

File: src/veil/ringct/lightwallet.cpp [hunk 14]
```diff
@@ -0,0 +1,1890 @@
...
+    } else {
+        errorMsg = "No Anon or Stealth txes given to build transaction";
+        return false;
+    }
+
+
+}
+
+
+/** Before you build the transaction, you must of already called
```
Comment (Zannick) [inline src/veil/ringct/lightwallet.cpp:149]: It should be reasonable to perform these safety checks here anyway. And, as I mentioned in review comments on the docs, sometimes it's better to perform that work here instead of expecting it to be done.

File: src/veil/ringct/lightwallet.cpp [hunk 15]
```diff
@@ -0,0 +1,1890 @@
...
+            vSpendTheseTx.emplace_back(tx);
+            if (currentSelected > nValueOut) {
+                tempmultipleamountchange = currentSelected - nValueOut;
+                fMultipleInput = true;
+                break;
+            }
+        }
+    }
+
+    LogPrintf("nValueOut %d, ", nValueOut);
```
Comment (Zannick) [inline src/veil/ringct/lightwallet.cpp:1333]:
```suggestion
```

File: src/veil/ringct/lightwallet.cpp [hunk 16]
```diff
@@ -0,0 +1,1890 @@
...
+
+    vecSend.insert(vecSend.begin() + nChangePositionOut, recipient);
+
+    return true;
+}
+
+bool SelectSpendableTxForValue(std::vector<CWatchOnlyTx>& vSpendTheseTx, CAmount& nChange, const CAmount& nValueOut, const std::vector<CWatchOnlyTx>& vSpendableTx)
+{
+    CAmount currentMinimumChange = 0;
+    CAmount tempsingleamountchange = 0;
```
Comment (Zannick) [inline src/veil/ringct/lightwallet.cpp:1296]: Please continue using camelCaseStyle

File: src/veil/ringct/lightwallet.cpp [hunk 17]
```diff
@@ -0,0 +1,1890 @@
...
+ * Light Node - Check you have at least the send amount + 0.01 Veil as a fee
+ * API - Fetch Valid Dummy Outputs
+ */
+// Args include, the following
+// Spend Secret
+// Scan Secret
+// Spend Public Key
+// Address to spend to
+// Amount to spend
+bool BuildLightWalletRingCTTransaction(const std::vector<std::string>& args, const std::vector<CWatchOnlyTx>& vSpendableTx, const std::vector<CLightWalletAnonOutputData>& vDummyOutputs, std::string& txHex, std::string& errorMsg)
```
Comment (Zannick) [inline src/veil/ringct/lightwallet.cpp:163]: This, plus the rest of the file, looks like a ton of copied code just to add in something lightwallet-related to each function? Is there a way we can deal with that directly in the original code instead or otherwise share functionality?

File: doc/lightwallet.md [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+1. A watchonly server setup and running, this server acts as a full Veil node syncing blocks and handling transaction information.
+   1. To run a server like this. All you need to do is run Veild with the flag -watchonly=1 placed in the veil.conf
+   2. Then you can setup a api server to act as the passthrough to the veild via rpc commands
+2. On your local desktop, you must have the ability to launch veild with the flag -lightwallet=1. This flag makes veild act like a light wallet. It will not connect to peers, it will not sync the blockchain. 
+   1. This is used to compute ringct transactions, to be broadcasted later via the watchonly server.
+---------------------------   
+## Prerequisites
+1. You have a ringct wallet that has a balance
+2. You have the ability to generate the following keys for the given address
+   1. Spend Secret Key
```
Comment (Zannick) [inline doc/lightwallet.md:17]: Needs instructions on how to generate these. Or they should be retrievable in the lightwallet itself so the user doesn't have to grab all these themselves (ideally the commands shouldn't have to prompt for them in that case, only check that the wallet is unlocked).

File: src/veil/ringct/watchonly.h [hunk 19]
```diff
@@ -0,0 +1,158 @@
+// Copyright (c) 2021 The Flux Core developers
```
Comment (Zannick) [inline src/veil/ringct/watchonly.h:1]: Please also add a line for Veil.

File: src/veil/ringct/watchonly.h [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+#include <key.h>
+#include <primitives/transaction.h>
+
+/** Map of watchonly keys */
+extern std::map<std::string, CWatchOnlyAddress> mapWatchOnlyAddresses;
+
+/** Global variable that points to the active watchonly database (protected by cs_main) */
+extern std::unique_ptr<CWatchOnlyDB> pwatchonlyDB;
+
+//class CWatchOnlyTX
```
Comment (Zannick) [inline src/veil/ringct/watchonly.h:31]: Remove this if it's dead.

File: src/veil/ringct/watchonlydb.cpp [hunk 21]
```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2021 The Flux Core developers
```
Comment (Zannick) [inline src/veil/ringct/watchonlydb.cpp:1]: Add a line for Veil.

File: src/veil/ringct/watchonlydb.h [hunk 22]
```diff
@@ -0,0 +1,58 @@
+// Copyright (c) 2021 The Flux Core developers
```
Comment (Zannick) [inline src/veil/ringct/watchonlydb.h:1]: Add a line for Veil.

**Comment by blondfrogs** [comment]
@Zannick Was there any issues related to the code or just the documentation of the code?

**Review by codeofalltrades — APPROVED** [review]
utACK 317f3131cfeeaa89e24ac106cabda4c7ba5d54c1

**Comment by Zannick** [comment]
I don't know what happened to my other comment, maybe I changed my mind and removed it. My opinion on the code was that it seemed much like copied and modified code from elsewhere and so it was hard to follow what the new elements were; I was not able to finish the code review at that time. But I do not want to block this review on a refactor. So, the majority of comments for this pass are on documentation, and then in a couple of days I will make another pass.

**Review by Zannick — APPROVED** [review]
utACK 317f313

I'd still like to see improvements to the documentation before release, but I'm okay with a separate PR for that.

**Comment by seanPhill** [comment]
I'll take a day or so at least of testing, and _maybe_ that will enable me to clarify the instructions a bit.

**Comment by seanPhill** [comment]
@blondfrogs What can I do to test this? Can we use a Zelcore wallet in testnet mode?

I have run in lightwallet mode, but not with every required component.

**Comment by seanPhill** [comment]
Not waiting for any further reviews. I've been running this on numerous active wallets for some time, so it at least doesn't break normal wallet operation. Time for Zelcore wallet.
